chapter_1/kmeans_1pluseps_approx.tex

```diff
@@ -0,0 +1,207 @@
+\end{lemma}
+
+\begin{proof}
+Let $\Pi = (S_1, S_2)$ and $\Pi' = (S_1', S_2')$. Given the optimality of Voronoi
+  clusters for given centroids, we need to show that the cost of a clustering
+  given $\epsilon$-near centroids is $\epsilon$-approximate, i.e. $$cost(S_1', c_1)
+  + cost(S_2' , c_2) \leq (1 + 6\epsilon)[cost(S_1, c_1) + cost(S_2 , c_2)]$$ This
+  is the same as showing for each $x \in S_1' \backslash S_1$ we have $\norm{x-c_1}
+  \leq (1+16\epsilon)\norm{x-c_2}$ and symmetric argument for $x \in S_2' \backslash S_2$.
+  Since $x \in S_1'$, x must be closer to $c_1'$ than $c_2'$, $\norm{x-c_1'}\leq  \norm{x-c_2'}$.
+  Let $\delta = \norm{c_1'-c_2'}$. By the triangle inequality and definition of $\epsilon$-near we have
+\begin{align*}
+    \norm{x-c_1} &  \leq \norm{x-c_1'} + \norm{c_1-c_1'} \\
+    & \leq \norm{x-c_1'} + \epsilon \norm{c_1'-c_2'}\\
+    & = \norm{x-c_1'} + \epsilon \delta \\
+    & \leq \norm{x-c_2'}+ \epsilon \delta \\
+    & \leq \norm{x-c_2} + \norm{c_2-c_2'} + \epsilon \delta \\
+    & \leq \norm{x-c_2} + 2\epsilon \delta
+\end{align*}
+we also have by similar reasoning
+$$\norm{x-c_2}\leq \norm{x-c_2'} - \epsilon \delta \leq \delta/2 - \epsilon \delta$$
+which implies $\delta \leq \frac{2}{1-2\epsilon}\norm{x-c_2}\leq 3 \norm{x-c_2}$. Using the bound on $\norm{x-c_2}$ and $\delta$, we have $\norm{x-c_1} \leq (1+ 6\epsilon)\norm{x-c_2}$ as required. 
+\end{proof}
+
+% We have already shown that for an $\epsilon$-approximate centroid set, the cost of a Voronoi partition for some pair of centroids is at most $(1+\epsilon)cost(\Pi)$ where $\Pi$ is any $2$-clustering. As the algorithm performs an exhaustive search of all pairs of centroids in the computed $\epsilon$-approximate centroid set, it is clear that it will thus try the least cost pairing and thus satisifes the $(1+\epsilon)$ approximation requirement.
+
+It remains to show that this algorithm can be performed in near-linear time. In a direct implementation of this algorithm, we must consider $|C|^2$ pairs $(c_1,c_2)$ and for each pair, we need $O(n)$ time for computing the cost of the corresponding $2$-clustering. The set $|C|$ has size $O(n \epsilon^{-d} \text{log}(1/\epsilon))$ thus implying $O(n^2)$ for a fixed $\epsilon$. However, we only need to consider an $\epsilon$-complete set of pairs of $P$ which has size $|P|=O(n \epsilon^{-d})=O(n)$.
+
+Further, using approximate range searching, computing the approximate cost can be done in $O(\text{log }n)$ time (down from $O(n)$). As each preprocessing step takes at most $O(n \log n)$ and the main for loop takes $O(n \log n)$ for a fixed $\epsilon$, the algorithm can thus run in $O(n \log n)$.
+
+The full k-clustering algorithm builds on the intuition of the 2-clustering version and employs a recursive approach for finding the $(1+\epsilon)$ approximate.
```

Comment:
apprxoimate what? There is a word missing or something at the end of this sentence